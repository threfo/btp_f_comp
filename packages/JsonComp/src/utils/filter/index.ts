import filterFuncMap from './filterFuncMap'

export const initFilter = (registerFilter: any): void => {
  Object.keys(filterFuncMap).forEach(key => {
    registerFilter(key, filterFuncMap[key])
  })
}
