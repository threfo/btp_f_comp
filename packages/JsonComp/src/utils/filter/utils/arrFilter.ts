import isFunction from 'lodash/isFunction'
import { getStrOrVariable } from '../helper'

type ModelType = 'notMatch' | 'match'

type RuleModelFunc = (input: any[], arr: any | any[]) => any[]
type RuleModelMap = {
  [key in ModelType]?: RuleModelFunc
}

export const notMatchFn = (input: any[], arr: any | any[]) => {
  if (Array.isArray(input)) {
    if (Array.isArray(arr)) {
      return input.filter(item => {
        return !arr.includes(item)
      })
    }
    return input.filter(item => item !== arr)
  }
  return []
}

export const matchFn = (input: any[], arr: any | any[]) => {
  if (Array.isArray(input)) {
    if (Array.isArray(arr)) {
      return input.filter(item => {
        return !!arr.find(arrItem => item.includes(arrItem))
      })
    }

    return input.filter(item => {
      return item.includes(arr || '')
    })
  }
  return []
}

const getModelFuncMap = (): RuleModelMap => ({
  notMatch: notMatchFn,
  match: matchFn
})

export const fn = (input: any[], arr: any | any[], model: ModelType) => {
  const func = getModelFuncMap()[model]

  let returnArr
  if (isFunction(func)) {
    returnArr = func(input, arr)
  }
  return returnArr || []
}

const filter: any = {
  fn(input: any[], arg: string, model?: ModelType) {
    const arr = getStrOrVariable(arg, this?.data, this?.filter?.args[2])

    return fn(input, arr, model || 'notMatch')
  }
}

export default filter.fn
