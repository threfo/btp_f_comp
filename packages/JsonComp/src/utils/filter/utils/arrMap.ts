import isString from 'lodash/isString'
import isObject from 'lodash/isObject'
import get from 'lodash/get'
import set from 'lodash/set'

import { resolveVariableAndFilter } from 'amis-formula'
import { createObject } from '../../index'

type ModelType = 'join' | 'map'

interface RuleItem {
  model?: ModelType // 解析策略 默认是 map
  key: string
  valueKeyMap?: any // key 获取到的值的映射map
  otherProps?: any
}
interface RuleMapItem {
  key?: string // 在新对象中的 key
  rule?: string | RuleItem // 字符串时为amis-formula filter的用法
}

interface RuleMap {
  [key: string]: RuleMapItem // 旧对象中的key: 对应的解析策略
}
interface ArrMapFilterInput {
  data: any
  ruleMap: RuleMap
  otherProps?: any
}

type RuleModelFunc = ({
  rule,
  data,
  value,
  obj,
  key
}: {
  rule: RuleItem | string
  data: any
  value: any
  obj: any
  key: string
}) => any

interface RuleModelMap {
  [key: string]: RuleModelFunc
}

export const valueKeyMapModel: RuleModelFunc = ({ rule, data, value }) => {
  // ruleKey 优先使用该 key 的值来操作
  const { key: ruleKey, valueKeyMap } = rule as RuleItem
  let ruleKeyValue
  if (ruleKey) {
    ruleKeyValue = get(data, ruleKey)

    if (valueKeyMap) {
      ruleKeyValue = valueKeyMap[ruleKeyValue]
    }
  }

  return ruleKeyValue || value
}

export const joinModel: RuleModelFunc = ({ rule, value, obj, key }) => {
  const { key: objKey, otherProps } = rule as RuleItem
  const arr = get(obj, key, [])

  if (value === undefined) {
    return arr
  }

  let valObj = value

  if (objKey) {
    valObj = {}
    set(valObj, objKey, value)

    if (otherProps) {
      valObj = {
        ...(otherProps || {}),
        ...valObj
      }
    }
  }

  arr.push(valObj)

  return arr
}

export const getRuleModelMap = (): RuleModelMap => ({
  map: valueKeyMapModel,
  join: joinModel
})

export const fn = (input: ArrMapFilterInput, showData: any) => {
  // console.log('registerFilter arrMap input:', input)

  const { data, ruleMap, otherProps } = input || {}

  if (!Array.isArray(data)) {
    console.warn('registerFilter arrMap input.data is not array', data)
    return []
  }

  const returnData = data.map((item: any) => {
    if (!isObject(item)) {
      if (isString(item) && isString(ruleMap)) {
        return {
          [ruleMap]: item
        }
      }

      return {}
    }

    return Object.keys(ruleMap || {}).reduce(
      (obj: any, ruleKey: string) => {
        const ruleItem = ruleMap[ruleKey]

        const { key = ruleKey, rule } = ruleItem

        const inputItem = item as any

        let value: any = get(inputItem, ruleKey)
        const contextData = createObject(showData, {
          v: value
        })
        if (rule) {
          if (isString(rule)) {
            // console.log('rule', rule)
            if (rule.includes('${')) {
              value = resolveVariableAndFilter(rule, contextData)
            } else {
              value = resolveVariableAndFilter(
                '${v' + ` | ${rule} ` + '}',
                contextData
              )
            }
          } else if (!Array.isArray(rule)) {
            const { model = 'map' } = rule

            const func = getRuleModelMap()[model]

            if (func) {
              value = func({
                rule,
                data: inputItem,
                value,
                obj,
                key
              })
            }
          }
        }

        set(obj, key, value)

        return obj
      },
      {
        ...(otherProps || {})
      } as any
    )
  })
  // console.log('registerFilter arrMap returnData:', returnData)
  return returnData
}

const filter: any = {
  fn(input: any) {
    return fn(input, this?.data)
  }
}

export default filter.fn
