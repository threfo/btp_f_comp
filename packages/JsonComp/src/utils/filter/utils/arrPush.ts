import { getStrOrVariable } from '../helper'

export const fn = (input: any[] | any, arr: any[]) => {
  if (Array.isArray(arr)) {
    if (Array.isArray(input)) {
      return [...arr, ...input]
    }
    return [...arr, input]
  }
  return []
}

const filter: any = {
  fn(input: any, arg: string) {
    const arr = getStrOrVariable(arg, this?.data, this?.filter?.args[2])

    return fn(input, arr)
  }
}

export default filter.fn
