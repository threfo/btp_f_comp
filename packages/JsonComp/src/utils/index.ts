export const createObject = <T>(prototype: any, properties: T): T => {
  const createdObject: any = Object.create(prototype || {})
  return Object.assign(createdObject, properties || {})
}

export const isPromise = (obj: any) =>
  !!obj &&
  (typeof obj === 'object' || typeof obj === 'function') &&
  typeof obj.then === 'function'
