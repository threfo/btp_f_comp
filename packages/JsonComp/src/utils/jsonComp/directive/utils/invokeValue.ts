import set from 'lodash/set'
import get from 'lodash/get'
import invoke from 'lodash/invoke'
import isString from 'lodash/isString'

import isFunction from 'lodash/isFunction'
import { DoDirectiveProps } from '../type'

import { resolve, resolveObj } from '../../../resole'
import { createObject, isPromise } from '../../../index'

type StepItemTargetType = 'that' | 'window' | 'actionProps' | 'invokeResult'
interface StepItem {
  key?: string // 在执行命令的目标上需要执行的路径
  isReturn?: string // 跳出执行的条件
  value?: any // 执行时的参数
  varName?: string // 定义后会以这个值作为上下文的声明
  defaultValue?: any // 获取指的时候的默认值
  target?: StepItemTargetType // 执行命令的目标，没有声明时为当前组件的this
}

export const checkIsReturn = (stepItem: StepItem, contextData: any) => {
  const { isReturn } = stepItem
  // console.log('invokeValue isReturn', isReturn)
  const resolveIsReturn = false
  if (isReturn !== undefined) {
    const resolveIsReturn = resolve(isReturn, contextData)
    // console.log('invokeValue resolveIsReturn', resolveIsReturn)
    if (resolveIsReturn) {
      return
    }
  }

  return resolveIsReturn
}

export const getInvokeTarget = ({
  that,
  stepItem,
  actionProps,
  invokeResultMap
}: {
  that: any
  stepItem: StepItem
  actionProps: any
  invokeResultMap: any
}) => {
  let invokeTarget = that
  const { target } = stepItem || {}
  if (target === 'window') {
    invokeTarget = window
  } else if (target === 'actionProps') {
    invokeTarget = actionProps
  } else if (target === 'invokeResult') {
    invokeTarget = invokeResultMap
  }

  return invokeTarget
}

export const invokeValue = async (props: DoDirectiveProps) => {
  const { that, directive, props: p } = props
  const { value: directiveValue } = directive
  const { steps } = directiveValue || {}

  // console.log('invokeValue', directive, props)
  // console.log('invokeValue that', that)
  // 缓存上一步的执行结构，给下一步可以调用
  const invokeResultMap: any = {}

  if (that && steps !== undefined) {
    const resoleData = createObject(that.__showData, { actionProps: p })
    // console.log('invokeValue resoleData', resoleData)

    let arr: StepItem[] = steps

    if (isString(steps)) {
      arr = resolve(steps, resoleData)
    }

    // console.log('invokeValue arr', arr)

    if (Array.isArray(arr)) {
      for (let i = 0; i < arr.length; i++) {
        const stepItem = arr[i] || {}
        const { key, value, varName, defaultValue } = stepItem
        // console.log('invokeValue stepItem', stepItem)

        const contextData = createObject(resoleData, {
          invokeResult: invokeResultMap
        })
        // console.log('invokeValue contextData', contextData)

        if (checkIsReturn(stepItem, contextData)) {
          // console.log('invokeValue isReturn')
          return
        }

        const invokeTarget = getInvokeTarget({
          that,
          stepItem,
          actionProps: p,
          invokeResultMap
        })

        const resolveKey = resolve(key || '', contextData)
        let invokeResult: any = get(invokeTarget, resolveKey)
        if (invokeResult === undefined) {
          invokeResult = defaultValue
        }

        // console.log('invokeValue resolveKey', resolveKey)
        // console.log('invokeValue invokeResult 1', invokeResult)
        let resolveValue
        if (value !== undefined) {
          resolveValue = resolveObj({
            obj: value,
            data: contextData,
            deep: true
          })
        }

        // console.log('invokeValue resolveValue', resolveValue)
        // console.log('invokeValue that', that)

        if (isFunction(invokeResult)) {
          invokeResult = invoke(
            invokeTarget,
            resolveKey,
            ...(resolveValue || [])
          )
          // console.log('invokeValue invokeResult is function')
          if (isPromise(invokeResult)) {
            // console.log('invokeValue invokeResult is Promise')
            invokeResult = await invokeResult

            // console.log('invokeValue await invokeResult:', invokeResult)
          }
        } else if (value !== undefined && resolveKey !== undefined) {
          const compDataKey = 'compData.'

          if (resolveKey.indexOf(compDataKey) === 0) {
            if (that.compData === undefined) {
              that.compData = {}
            }
          }

          if (resolveKey.indexOf(compDataKey) === 0 && !that.res) {
            // console.log('invokeValue Vue.set')
            const willSetKey = resolveKey.replace(compDataKey, '')
            if (that.$set) {
              that.$set(that.compData, willSetKey, resolveValue)
            } else {
              that.compData[willSetKey] = resolveValue
            }
          } else {
            // console.log(
            //   'invokeValue set',
            //   invokeTarget,
            //   resolveKey,
            //   resolveValue
            // )
            set(invokeTarget, resolveKey, resolveValue)
          }
        }

        // console.log('invokeValue invokeResult 2', invokeResult)

        const invokeResultMapKey = varName
        if (invokeResultMapKey && !isFunction(invokeResult)) {
          invokeResultMap[invokeResultMapKey] = invokeResult
        }
      }
    }
  }

  // console.log('invokeValue that.compData', that.compData)

  return await Promise.resolve(invokeResultMap)
}

export default async () => {
  return await Promise.resolve(invokeValue)
}
