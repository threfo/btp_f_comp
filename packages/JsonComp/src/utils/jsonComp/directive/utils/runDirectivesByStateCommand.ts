import isFunction from 'lodash/isFunction'
import { DoDirectiveProps } from '../type'

import { resolveObj } from '../../../resole'
import { createObject } from '../../../index'

export const runDirectivesByStateCommand = async (props: DoDirectiveProps) => {
  const { that, directive, props: p } = props

  const { setStateCommand } = that || {}
  // console.log('runDirectivesByStateCommand', directive, props)
  if (!isFunction(setStateCommand)) {
    console.error('this.setStateCommand not exist or not a function')
    return
  }

  const resoleData = createObject(that.__showData, { actionProps: p })

  const value = resolveObj({
    obj: directive.value || {},
    data: resoleData,
    deep: true
  })

  const { value: directiveList } = value || {}

  if (!directiveList) {
    console.error('value.value not exist')
    return
  }

  // console.log('runDirectivesByStateCommand value:', value)

  that.setStateCommand({
    ...value,
    type: 'runCommandDirectives'
  })

  return await Promise.resolve(that.compData)
}

export default async () => {
  return await Promise.resolve(runDirectivesByStateCommand)
}
