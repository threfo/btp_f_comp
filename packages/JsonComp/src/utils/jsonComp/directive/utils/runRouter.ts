import isFunction from 'lodash/isFunction'
import isString from 'lodash/isString'
import { DoDirectiveProps } from '../type'

import { resolveObj, resolve } from '../../../resole'
import { createObject } from '../../..'

export const runRouter = async (props: DoDirectiveProps) => {
  const { that, directive, props: p } = props
  const { value } = directive
  console.log('runRouter', directive, p)
  if (that && value !== undefined) {
    const { path: valuePath, method: valueMethod = 'push' } = value

    let { query: valueQuery } = value

    const { query: routeQuery, path: routerPath } = that.$route || {}

    const resoleData = createObject(that.__showData, { actionProps: p })

    // console.log('valueQuery', valueQuery)
    // console.log('routeQuery', routeQuery)
    console.log('runRouter resoleData', resoleData)

    if (isString(valueQuery)) {
      valueQuery = resolve(valueQuery, resoleData)
    }

    let query = resolveObj({
      obj: {
        ...(routeQuery || {}),
        ...(valueQuery || {})
      },
      data: resoleData,
      deep: true
    })

    const path = resolve(valuePath || routerPath, resoleData)

    const method = resolve(valueMethod, resoleData)

    console.log('runRouter query', query)
    // console.log('path', path)
    // console.log('method', method)

    if (query) {
      query = Object.keys(query).reduce((obj: any, key: string) => {
        const val = query[key]
        if (val !== null) {
          obj[key] = val
        }

        return obj
      }, {})
    }

    const func = that.$router[method]
    if (isFunction(func)) {
      that.$router[method]({
        path,
        query
      })
    } else {
      console.error(`runRouter method: ${method} not a function`)
    }
  }

  return await Promise.resolve(that.compData)
}

export default async () => {
  return await Promise.resolve(runRouter)
}
