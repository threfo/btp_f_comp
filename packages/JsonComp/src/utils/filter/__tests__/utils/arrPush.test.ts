import { fn } from '../../utils/arrPush'
describe('utils/filter/utils/arrPush.ts', () => {
  it('fn', () => {
    expect(fn('2', ['1', '2'])).toMatchSnapshot()

    expect(fn(['1', '2', '3'], ['2', '3', '4'])).toMatchSnapshot()
  })
})
