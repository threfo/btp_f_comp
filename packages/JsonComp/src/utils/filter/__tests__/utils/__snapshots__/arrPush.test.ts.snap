// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/filter/utils/arrPush.ts fn 1`] = `
Array [
  "1",
  "2",
  "2",
]
`;

exports[`utils/filter/utils/arrPush.ts fn 2`] = `
Array [
  "2",
  "3",
  "4",
  "1",
  "2",
  "3",
]
`;
